src/circuit_builder.rs

```diff
@@ -38,44 +273,162 @@ impl PublicInput {
+    fn build_pi(&self, pub_inputs: &[PublicInput]) -> Vec<BlsScalar> {
+        let mut pi = vec![BlsScalar::zero(); self.circuit_size()];
+
+        pub_inputs.iter().for_each(|p| match p {
+            PublicInput::BlsScalar(s, pos) => pi[*pos] = *s,
+            PublicInput::JubJubScalar(s, pos) => pi[*pos] = BlsScalar::from(*s),
+            PublicInput::AffinePoint(p, pos_x, pos_y) => {
+                pi[*pos_x] = p.get_x();
+                pi[*pos_y] = p.get_y();
+            }
+        });
+
+        pi
+    }
 
     /// Get the circuit size of the implemented circuit.
     fn circuit_size(&self) -> usize;
 
+    /// Set the circuit size of the implemented circuit.
+    fn set_circuit_size(&mut self, circuit_size: usize);
+
+    /// Generates the public inputs for a verifier.
+    fn gen_public_inputs(&mut self) -> Result<Vec<PublicInput>, Error> {
+        // TODO - Create a dummy composer
+        let mut verifier = Verifier::new(Self::transcript_initializer());
+
+        self.gadget(verifier.mut_cs())
+    }
+
     /// Generates a proof using the provided `CircuitInputs` & `ProverKey` instances.
-    fn gen_proof(
-        &mut self,
-        pub_params: &PublicParameters,
-        prover_key: &ProverKey,
-        transcript_initialisation: &'static [u8],
-    ) -> Result<Proof>;
+    fn gen_proof(&mut self) -> Result<Proof, Error> {
+        let mut prover = Prover::new(Self::transcript_initializer());
+
+        self.gadget(prover.mut_cs())?;
+
+        prover.prover_key = self.prover_key().as_ref().cloned();
+        self.commit_key()
+            .as_ref()
+            .ok_or(anyhow!("Commit key not generated"))
+            .and_then(|ck| prover.prove(ck))
+    }
 
     /// Verifies a proof using the provided `CircuitInputs` & `VerifierKey` instances.
-    fn verify_proof(
-        &mut self,
-        pub_params: &PublicParameters,
-        verifier_key: &VerifierKey,
-        transcript_initialisation: &'static [u8],
-        proof: &Proof,
-        pub_inputs: &[PublicInput],
-    ) -> Result<()>;
+    fn verify_proof(&mut self, proof: &Proof, pub_inputs: &[PublicInput]) -> Result<(), Error> {
+        let pi = self.build_pi(pub_inputs);
+
+        let mut verifier = Verifier::new(Self::transcript_initializer());
+
+        verifier.verifier_key = self.verifier_key().as_ref().cloned();
+        self.opening_key()
+            .as_ref()
+            .ok_or(anyhow!("Opening key not generated"))
+            .and_then(|ok| verifier.verify(proof, ok, pi.as_slice()))
+    }
+
+    /// Commit key generated by `compile`
+    fn commit_key(&self) -> &Option<CommitKey>;
+
+    /// Setter for commit key generated by `compile`
+    fn set_commit_key(&mut self, ck: CommitKey);
+
+    /// Opening key generated by `compile`
+    fn opening_key(&self) -> &Option<OpeningKey>;
+
+    /// Setter for opening key generated by `compile`
+    fn set_opening_key(&mut self, ok: OpeningKey);
```

Comment:
Not worth it to have imo. Makes the API bigger and more verbose than just using the PubParams directly.

Reply:
These are all boilerplate functions with a default implementation each